File: retrieve_uniprot_proteomes.pl
#!/usr/bin/env perl

# Adapted from UniProt script (http://www.uniprot.org/help/programmatic_access)

use strict;
use warnings;
use LWP::UserAgent;
use HTTP::Date;

# Taxonomy identifier of top node for query, e.g. 2 for Bacteria, 2157 for Archea, etc.
# (see http://www.uniprot.org/taxonomy)
my $top_node = $ARGV[0];

my $reference = 0; # Toggle this to 1 if you want reference instead of complete proteomes.
my $proteome = $reference ? 'reference:yes' : 'complete:yes';
my $keyword = $reference ? 'keyword:1185' : 'keyword:181';

my $contact = ''; # Please set your email address here to help us debug in case of problems.
my $agent = LWP::UserAgent->new(agent => "libwww-perl $contact");

# Get a list of all taxons below the top node with a complete/reference proteome.
my $query_list = "http://www.uniprot.org/taxonomy/?query=ancestor:$top_node+$proteome&format=list";
my $response_list = $agent->get($query_list);
die 'Failed, got ' . $response_list->status_line .
  ' for ' . $response_list->request->uri . "\n"
  unless $response_list->is_success;

# For each taxon, mirror its proteome in FASTA format.
for my $taxon (split(/\n/, $response_list->content)) {
  my $file = $taxon . '.fasta';
  my $query_taxon = "http://www.uniprot.org/uniprot/?query=organism:$taxon+$keyword&format=fasta&include=yes";
  my $response_taxon = $agent->mirror($query_taxon, $file);

  if ($response_taxon->is_success) {
    my $results = $response_taxon->header('X-Total-Results');
    my $release = $response_taxon->header('X-UniProt-Release');
    my $date = sprintf("%4d-%02d-%02d", HTTP::Date::parse_date($response_taxon->header('Last-Modified')));
    print "File $file: downloaded $results entries of UniProt release $release ($date)\n";
  }
  elsif ($response_taxon->code == HTTP::Status::RC_NOT_MODIFIED) {
    print "File $file: up-to-date\n";
  }
  else {
    die 'Failed, got ' . $response_taxon->status_line .
      ' for ' . $response_taxon->request->uri . "\n";
  }
}

File: retrieve_uniprot_proteomes.py
#!/usr/bin/env python3

import argparse
import httplib2
import sys

__version__ = '0.0.0'
__prog__ = 'retrieve_uniprot_proteomes'

LIST_TEMPLATE = "http://www.uniprot.org/taxonomy/?query=ancestor:{}+{}&format=list"
RET_TEMPLATE  = "http://www.uniprot.org/uniprot/?query=organism:{}+{}&format=fasta&include=yes"

def positive_int(i):
    i = int(i)
    if i < 0:
         raise argparse.ArgumentTypeError("%s is an invalid positive number" % i)
    return i

def parser(argv=None):
    parser = argparse.ArgumentParser()
    parser.add_argument(
        '--version',
        help='Display version',
        action='version',
        version='%(prog)s {}'.format(__version__)
    )
    parser.add_argument(
        'top_taxid',
        type=positive_int,
        help='The taxon id of the ancestral node'
    )
    parser.add_argument(
        '-r', "--reference",
        help="retrieve only reference proteomes",
        action="store_true",
        default=False
    )
    parser.add_argument(
        '-d', '--download_ids',
        help="download the list of ids",
        action="store_true",
        default=False
    )
    parser.add_argument(
        '--print_http',
        help="Print all HTTP request",
        action="store_true",
        default=False
    )
    parser.add_argument(
        '-c', '--cache',
        help="Cache directory name",
        default="uniprot-cache"
    )
    args = parser.parse_args(argv)
    return(args)

def prettyprint_http(response):
    d = dict(response.items())
    for k,v in d.items():
        print("\t%s: %s" % (k,v), file=sys.stderr)


if __name__ == '__main__':

    args = parser()

    if args.print_http:
        httplib2.debuglevel = 1

    proteome = 'reference:yes' if args.reference else 'complete:yes'
    keyword = 'keyword:1185' if args.reference else 'keyword:181'

    h = httplib2.Http(args.cache)

    list_url = LIST_TEMPLATE.format(args.top_taxid, proteome)
    response, content = h.request(list_url)
    if args.print_http:
        prettyprint_http(response)
    taxids = content.decode().strip().split("\n")

    for taxid in taxids:
        if args.download_ids:
            filename = '%s.faa' % taxid
            print("Retrieving %s" % filename, file=sys.stderr)
            url = RET_TEMPLATE.format(taxid, keyword)
            response, content = h.request(url)
            with open(filename, 'w') as f:
                print(content.decode(), file=f)
        else:
            print(taxid)
